### PR TITLE
Implement LRU Cache

### PR DESCRIPTION
**Pull request by Nama21yo**
## Description
This PR introduces **LRU caching** into the incremental feature selection workflow and refactors several functions to use **pattern matching overloads** instead of `unify`.

### Key changes:
- Implemented `lruCache` and applied it to incremental selection
- Replaced `unify`-based branching with **function overloading** for:
- Added **structured tests** for caching behavior, eviction, overwrites, and edge cases.

## Motivation and Context
- **Performance:**  
  Previously, incremental feature selection took **~2 minutes 7 seconds on average**.  
image
image

  With caching, the runtime has been reduced to **~33.7 seconds on average**.
- **Readability & Maintainability:**  
  Using pattern matching instead of `unify` makes code more idiomatic, cleaner, and easier to reason about.

## How Has This Been Tested?
- Added unit tests for:
  - `cachedFactorial` (hits, misses, eviction, overwrite)
  - `cachedMultiply` (hits, misses, edge cases with zero)
  - `cachedMutualInformation` (cache behavior across multiple calls)
- Verified **LRU eviction** and **cache size consistency**.
- Confirmed feature selection workflow runs significantly faster with caching.
## Types of changes
- [x] Performance improvement (non-breaking change that improves efficiency)
- [x] Refactor (cleaner implementation using pattern matching instead of unify)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] Performance benchmarks recorded.

### CODE REVIEW

File: utilities/lru-cache.metta [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+        (
+            ; (() (println! (DEBUG Inside cacheExists $functionName $argument)))
+            ($queryResult (collapse (match &cache (cacheEntry $functionName $argument $result $timestamp) $result)))
+            ; (() (println! (DEBUG query Result: $queryResult)))
+        )
+        (not (== $queryResult ()))
+    )
+)
+
+;; Function: nextTimestamp
```
Comment (K1rubel) [inline utilities/lru-cache.metta:32]: `nextTimestamp`. This is not an actual timestamp, it is a counter, isn't it? Better if you rename this appropriately.

File: utilities/lru-cache.metta [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+;; Purpose: Evict the least recently used cache entry to make room for new entry
+(: removeLeastRecentlyUsed (-> Symbol Any Any Any))
+(= (removeLeastRecentlyUsed $funcs $arguments $results) 
+    (let* 
+        (
+            ;; Get all access time entries
+            ($allAccessTimes (collapse (match &cache (accessTime $fn $args $timestamp) (accessEntry $fn $args $timestamp))))
+        )
+        (if (== $allAccessTimes ())
+            (Error "Cache Empty")
```
Comment (K1rubel) [inline utilities/lru-cache.metta:102]: `Error` requires two atoms. Correct it to sth like this `(Error $atom $description)`

File: utilities/lru-cache.metta [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+        )
+        (if (>= $currentSize $maxSize)
+            ;; Cache is full, evict least recently used
+            (removeLeastRecentlyUsed $functionName $argument $result)
+            ;; Cache has space, add directly
+            (let* 
+                (
+                    ;; Add cache entry and access time
+                    (() (add-atom &cache (cacheEntry $functionName $argument $result $timestamp)))
+                    (() (add-atom &cache (accessTime $functionName $argument $timestamp)))
```
Comment (K1rubel) [inline utilities/lru-cache.metta:147]: ```146 and 147```
I feel like these entries could be done in one expression atom. I mean you already have the timestamp information in the `(cacheEntry ..)` expression atom and then you add the timestamp as a separate atom to the `&cache` space in the `(accessTime ... )` expression atom.

Comment (Nama21yo) [reply utilities/lru-cache.metta]: Yeah previously I thought I could implement time-to-live and related things in caching that was the point of separation came to me but on the current implementation Yeah you are right only `cacheEntry` is enough I have updated it in that manner.
And also I have removed all the unnecessary comments and debug statements.
And Yeah I have also changed the `timeStamp` to counter😅.

File: utilities/lru-cache.metta [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+;; State variables for cache management
+!(bind! cacheSize (new-state 0))         ;; Current number of cached items
+!(bind! maxCacheSize (new-state 10))     ;; Maximum cache capacity
+!(bind! accessCounter (new-state 0))     ;; sequential counter for ordering
+
+!(bind! DEFAULT_CACHE_SIZE 60)
+
+;; Function: cacheExists
+;; Purpose: Check if a function call with given arguments exists in cache
+(: cacheExists (-> Symbol Any Bool))
```
[the remark this answers is not in the document]
Comment (Yagth) [reply utilities/lru-cache.metta:18]: Resolve this conversation when done :)

**Comment by Nama21yo** [comment]
> Awesome implementation overall. Great work on generalizing the lru cache in a way it can be used for any of our functions.

Thanks @Yagth it means a lot.
